### PR TITLE
dragonball: metrics support for the upcall client

### PR DESCRIPTION
**Pull request by lisongqian**
This PR adds metrics support for the upcall client.

Fixes: #7248

Signed-off-by: Songqian Li <mail@lisongqian.cn>

### CODE REVIEW

File: src/dragonball/src/dbs_upcall/src/lib.rs [hunk 1]
```diff
@@ -356,21 +397,24 @@ impl<S: UpcallClientService + Send> UpcallEpollHandler<S> {
...
+                        warn!("upcall response failed {}", e);
+                        info.set_state(UpcallClientState::WaitingServer);
+                        if let Err(e) = self.set_reconnect() {
+                            error!("set reconnect error: {}", e);
+                            info.set_state(UpcallClientState::ReconnectError);
+                        }
                     }
                 }
-            },
+                self.metrics.handle_response_count.inc();
```
Comment (studychao) [inline src/dragonball/src/dbs_upcall/src/lib.rs:416]: I think the metric for how many response we handled in upcall is not very useful..

File: src/dragonball/src/dbs_upcall/src/lib.rs [hunk 2]
```diff
@@ -275,18 +303,29 @@ pub struct UpcallEpollHandler<S: UpcallClientService + Send> {
...
             in_reconnect: false,
+            metrics: metrics.clone(),
         };
         let info = handler.info.clone();
-        info.lock().unwrap().server_connection_start()?;
+        info.lock()
+            .unwrap()
+            .server_connection_start()
+            .map_err(|e| {
+                metrics.connect_start_fails.inc();
```
Comment (studychao) [inline src/dragonball/src/dbs_upcall/src/lib.rs:326]: there are multiple places using `server_connection_start()` but related metric only count here which is not enough.

How about add this metric inc() into the function `server_connection_start`?

File: src/dragonball/src/dbs_upcall/src/lib.rs [hunk 3]
```diff
@@ -316,6 +356,7 @@ impl<S: UpcallClientService + Send> UpcallEpollHandler<S> {
     }
 
     fn handle_stream_event(&mut self, ops: &mut EventOps) {
+        self.metrics.stream_event_count.inc();
```
Comment (studychao) [inline src/dragonball/src/dbs_upcall/src/lib.rs:359]: not very useful metric to me..

**Review by studychao — COMMENTED** [review]
thanks, a few comments.

File: src/dragonball/src/dbs_upcall/src/lib.rs (hunk 1 above)
Comment (lisongqian) [reply src/dragonball/src/dbs_upcall/src/lib.rs]: Okay, I removed it.

File: src/dragonball/src/dbs_upcall/src/lib.rs (hunk 3 above)
Comment (lisongqian) [reply src/dragonball/src/dbs_upcall/src/lib.rs]: I removed it.

File: src/dragonball/src/dbs_upcall/src/lib.rs (hunk 2 above)
Comment (lisongqian) [reply src/dragonball/src/dbs_upcall/src/lib.rs]: Agree with you. I fixed it.

Comment (ZizhengBian) [reply src/dragonball/src/dbs_upcall/src/lib.rs]: I think the count of upcall request count is helpful, we can add this to metrics.

Comment (lisongqian) [reply src/dragonball/src/dbs_upcall/src/lib.rs]: The metric of upcall request counts was added to metrics.

**Review by ZizhengBian — APPROVED** [review]
LGTM

**Review by studychao — APPROVED** [review]
LGTM

**Pull request closed without merging**